audit_log/admin.py

```diff
@@ -0,0 +1,228 @@
+                request=request,
+                operation=Operation.READ.value,
+            )
+        )
+
+    def save_model(self, request, obj, form, change):
+        """
+        Given a model instance save it to the database.
+        """
+        # TODO: add a change log to the audit log message
+        message = create_api_commit_message_from_request(
+            request=request,
+            operation=Operation.UPDATE.value if change else Operation.CREATE.value,
+            object_ids=[str(obj.pk)],
+            model_name=obj._meta.model_name,
+        )
+        audit_log_service._commit_to_audit_log(message=message)
+        super().save_model(request, obj, form, change)
+
+    def delete_model(self, request, obj):
+        """
+        Given a model instance delete it from the database.
+        """
+        message = create_api_commit_message_from_request(
+            request=request,
+            operation=Operation.DELETE.value,
+            object_ids=[str(obj.pk)],
+            model_name=obj._meta.model_name,
+        )
+        audit_log_service._commit_to_audit_log(message=message)
+        super().delete_model(request, obj)
+
+    def delete_queryset(self, request, queryset):
+        """Given a queryset, delete it from the database."""
+        super().delete_queryset(
+            request,
+            queryset.with_audit_log_and_request(
+                request=request, operation=Operation.DELETE.value
+            ),
+        )
+
+
+class AuditLogMessageOperationListFilter(admin.SimpleListFilter):
+    # Human-readable title which will be displayed in the
+    # right admin sidebar just above the filter options.
+    title = _("operation")
+
+    # Parameter for the filter that will be used in the URL query.
+    parameter_name = "operation"
+
+    options = [operation.value for operation in Operation]
+
+    def lookups(self, request, model_admin):
+        return map(lambda o: (o, o.upper()), self.options)
+
+    def queryset(self, request, queryset):
+        if self.value() in self.options:
+            return queryset.filter(message__icontains=self.value())
+        return queryset
+
+
+class AuditLogMessageRoleListFilter(admin.SimpleListFilter):
+    # Human-readable title which will be displayed in the
+    # right admin sidebar just above the filter options.
+    title = _("role")
+
+    # Parameter for the filter that will be used in the URL query.
+    parameter_name = "role"
+
+    options = [role.value for role in Role]
+
+    def lookups(self, request, model_admin):
+        return map(lambda o: (o, o.upper()), self.options)
+
+    def queryset(self, request, queryset):
+        if self.value() in self.options:
+            return queryset.filter(message__icontains=self.value())
+        return queryset
+
+
+class LargeTablePaginator(Paginator):
+    """
+    Paginator that uses PostgreSQL `reltuples` for queryset size.
+
+    The reltuples values is just an estimate, but it's much faster than
```

Comment:
As the `reltuples` is just an estimate then doesn't that mean that the paginator can work incorrectly i.e. make it so that some pages are missed?

Reply:
So the bottom line is that this will not give precise paginations and will therefore result in wrong count / paginations in the audit log entries in the Django admin, but that's good enough as it's not really the place where audit log entries should be viewed, having them in the application database is meant for temporary storage until they are sent to ElasticSearch. Although saying this out loud it also sounds like what's the use of planning for millions of rows if there shouldn't be millions of rows as they should be pushed to ElasticSearch unless there are a HUGE amount of audit logs coming each day. But yeah, I'm still not 100% ok with this, but I won't stall this PR just because of this point.